feat(sync-call): [IC-1666] Endpoint for synchronous call requests (#265)
dfinity/interface-spec: spec/index.md

## HTTPS Interface {#http-interface}

The concrete mechanism that users use to send requests to the Internet Computer is via an HTTPS API, which exposes four endpoints to handle interactions, plus one for diagnostics:

-   At `/api/v2/canister/<effective_canister_id>/call` the user can submit update calls that are asynchronous and might change the IC state.

> -   At `/api/v2/canister/<effective_canister_id>/sync_call` the user can submit update calls and either get a synchronous HTTPS response with a certified status (if the update call completes within one execution round) or have the update call processed asynchronously.

Reviewer: ```suggestion
-   At `/api/v2/canister/<effective_canister_id>/sync_call` the user can submit update calls and either get a synchronous HTTPS response with a certified status or have the update call processed asynchronously.
```
I wouldn't specify those implementation details.

### Overview of canister calling {#http-call-overview}

Users interact with the Internet Computer by calling canisters. By the very nature of a blockchain protocol, they cannot be acted upon immediately, but only with a delay. Moreover, the actual node that the user talks to may not be honest or, for other reasons, may fail to get the request on the way.

The Internet Computer has two HTTPS APIs for canister calling:

> - [*Asynchronous*](#http-async-call) canister calling, where the user must poll the Internet Computer for the status of the request.
> - [*Synchronous*](#http-sync-call) canister calling, where the user waits for a certified response from the Internet Computer for the initial call.

Reviewer: ```suggestion
- [*Asynchronous*](#http-async-call) canister calling, where the user must poll the Internet Computer for the status of the canister call by _separate_ HTTPS requests.
- [*Synchronous*](#http-sync-call) canister calling, where the Internet Computer serves the response of the canister call as the response to the original HTTPS request.
```

When asking the IC about the state or call of a request, the user uses the request id (see [Request ids](#request-id)) to read the request status (see [Request status](#state-tree-request-status)) from the state tree (see [Request: Read state](#http-read-state)).

#### Synchronous canister calling {#http-sync-call}

> A synchronous update call, also known as a "call and await", is a type of update call where the replica will respond with a certificate if the call completes within one execution round.

Reviewer: ```suggestion
A synchronous update call, also known as a "call and await", is a type of update call where the replica will respond with a certificate if the call completes within an implementation-specific timeout.
```

The purpose of the synchronous call endpoint is to give the user a certified response for their canister call. This means that if a certificate is returned, then the user __does not need to poll__ (using [`read_state`](#http-read-state) requests) to determine the result of the call.

A certificate is returned by the IC to the user if:

> -   The execution of the call completes within one execution round, and the status is certified by the IC within the ingress expiry time. The status will be either `replied`, `rejected` or `done`.

Reviewer: ```suggestion
-   The execution of the call completes within an implementation-specific timeout. The status will be either `replied`, `rejected` or `done`.
```
Note that returning a certificate with `done` doesn't really reveal what happened.

> -   The call expires before it is queued and is dropped by the IC. The status of the call will be `unknown` with the certificate time stamp at a later time than the ingress expiry time.

Reviewer: ```suggestion
```

Reviewer: > The status of the call will be `unknown` with the certificate time stamp at a later time than the ingress expiry time.

Such a certificate isn't really useful.

The synchronous call endpoint is useful for users as it removes the networking overhead of polling the IC to determine the status of their call.

> Synchronous calls are therefore best suited for calls that are fast and expected to complete within one execution round. If a call does not complete within one execution round, the replica won't return a certificate and the user will need to poll the IC to determine the status of the call.

Reviewer: ```suggestion
Synchronous calls are therefore best suited for calls that are fast. If a call does not complete within an implementation-specific timeout, the replica won't return a certificate and the user will need to poll the IC to determine the status of the call.
```

When asking the IC about the state or call of a request, the user uses the request id (see [Request ids](#request-id)) to read the request status (see [Request status](#state-tree-request-status)) from the state tree (see [Request: Read state](#http-read-state)).

> #### Synchronous canister calling {#http-sync-call}

Reviewer: I don't get the point of this section - it looks like you copied the first eight steps from the previous section.

Reviewer: Instead of repeating the steps, I'd just add a few sentences describing a synchronous call as the replica making an asynchronous call followed by polling for its status on behalf of the caller (until a timeout).

Author: I removed the step-by-step points. Can you take another look?

> Unlike asynchronous canister calling, synchronous canister calling does not require the user to poll the Internet Computer for the status of the request. Instead, the user waits for a certified response from the Internet Computer for the initial call. However, The state transitions and semantics of the call are the same as for asynchronous canister calling.

Reviewer: ```suggestion
Unlike asynchronous canister calling, synchronous canister calling does not require the user to poll the Internet Computer for the status of the request. Instead, the user waits for a certified response from the Internet Computer for the initial call. Still, the state transitions and semantics of the call itself are the same as for asynchronous canister calling.
```

> Unlike asynchronous canister calling, synchronous canister calling does not initially require the user to poll the Internet Computer for the status of the request. Instead, the user waits for a certified response from the Internet Computer for the initial call. However, if a certified response is not received (due to a timeout or pre-execution error), the user must poll the Internet Computer for the status. Still, the state transitions and semantics of the call itself are the same as for asynchronous canister calling.

Reviewer: In case of a pre-execution error, polling makes no sense since the request status is not in the certified state tree (in such a case, the HTTP handler synchronously rejects the request).

Author: What if the replica is malicious? I.e. it sends a `reject` response back to the user, but actually submitted the ingress message.

Author: I rewrote the section and removed this part.

> Unlike asynchronous canister calling, synchronous canister calling does not initially require the user to poll the Internet Computer for the status of the request. Instead, the user waits for a certified response from the Internet Computer for the initial call. However, if a certified response is not received (due to a timeout or pre-execution error), the user must poll the Internet Computer for the status. Still, the state transitions and semantics of the call itself are the same as for asynchronous canister calling.

Reviewer: ```suggestion
Unlike asynchronous canister calling, synchronous canister calling does not initially require the user to poll the Internet Computer for the status of the request. Instead, the user waits for a certified response from the Internet Computer for the initial call. However, if a certified response is not received (due to a timeout), the user must poll the Internet Computer for the status of the request. In that case, the state transitions and semantics of the call itself are the same as for asynchronous canister calling.
```

Reviewer 2: I strongly believe that the replica should be allowed to conclude a synchronous call before it times out. E.g. if it is stuck in `processing` for too long. Or even IMHO if it's in the received` state for more than a minute or so (someone could attempt to DoS a single canister with synchronous calls; we don't want a replica maintaining 300k concurrent synchronous calls).

So maybe we should be more explicit about replicas being able to terminate a synchronous call at any point before the actual call has timed out or completed.

Could be as simple as replacing "if a certified response is not received (due to a timeout)" with "if a certified response is not produced in reasonable time".

Author: Agreed. It's more explicit about it now.

1.  A user submits a synchronous call via the [HTTPS Interface](#http-interface).

> 2.  The IC asks the targeted canister if it is willing to accept this message and be charged for the expense of processing it. This uses the [Ingress message inspection](#system-api-inspect-message) API for normal calls. For calls to the management canister, the rules in [The IC management canister](#ic-management-canister) apply.

Reviewer: ```suggestion
2.  The IC asks the targeted canister if it is willing to accept this message and be charged for the expense of processing it. This uses the [Ingress message inspection](#system-api-inspect-message) API for calls to regular canisters. For calls to the management canister, the rules in [The IC management canister](#ic-management-canister) apply.
```

> 3.  At some point, the IC may accept the call for processing and set its status to `received`. This indicates that the IC as a whole has received the call and plans on processing it (although it may still not get processed if the IC is under high load).

Reviewer: ```suggestion
3.  At some point, the IC may accept the call for processing and set its status to `received`. This indicates that the IC as a whole has received the call and plans on processing it (although the IC may still not start processing the request if it is under high load).
```

4.  If the call is processed (sufficient resources, call not yet expired), it will be executed, for some calls this may be atomic, for others this involves multiple internal steps.

5.  Eventually, a response will be produced which will be replied the user. The response can be a `reply`, indicating success, a `reject`, indicating some form of error.

> 6.  In the case that the call has been retained for long enough before a response is generated, but the request has not expired yet, the IC can forget the response data and only remember the call as `done`, to prevent a replay attack.

Author: Thanks. I tried to simplify the section. Could you take another look?

When asking the IC about the state or call of a request, the user uses the request id (see [Request ids](#request-id)) to read the request status (see [Request status](#state-tree-request-status)) from the state tree (see [Request: Read state](#http-read-state)).

#### Synchronous canister calling {#http-sync-call-overview}

> A synchronous update call, also known as a "call and await", is a type of update call where the replica will attempt to respond with a certificate of the state of the call to the HTTPS request. If the returned certificate indicates that the update call is in a terminal state (`replied`, `rejected`, or `done`), then the user __does not need to poll__ (using [`read_state`](#http-read-state) requests) to determine the result of the call. A terminal state means the call has completed its execution.

Reviewer: ```suggestion
A synchronous update call, also known as a "call and await", is a type of update call where the replica will attempt to respond to the HTTPS request with a certificate of the call status. If the returned certificate indicates that the update call is in a terminal state (`replied`, `rejected`, or `done`), then the user __does not need to poll__ (using [`read_state`](#http-read-state) requests) to determine the result of the call. A terminal state means the call has completed its execution.
```

The synchronous call endpoint is useful for users as it removes the networking overhead of polling the IC to determine the status of their call.

> The replica will maintain the HTTPS connection for the request and will respond once the state transitions to a terminal state. If an implementation specific timeout for the request is reached while the replica waits for the terminal state, then the replica will reply before the call completes, meaning a certificate for the call state in `unknown`, `received`, or `processing` is returned.

Reviewer: ```suggestion
The replica will maintain the HTTPS connection for the request and will respond once the call status transitions to a terminal state. If an implementation specific timeout for the request is reached while the replica waits for the terminal state, then the replica will reply before the call completes.
```

Author: See my comment [here](https://github.com/dfinity/interface-spec/pull/265#discussion_r1576248114). We should return certificates for the other states as well as it allows the user to know whether their ingress message made it to the IC.

Author: Changed it to return `202` and an empty body if it's not in a terminal state.

The user should take the following actions for the returned certificate states:

-    `processing`: The call is being executed by the IC. The user should poll the IC using [`read_state`](#http-read-state) requests to determine the result of the call.

-    `received`: The call has been received by the IC. The user should poll the IC using [`read_state`](#http-read-state) requests to determine the result of the call.

> -    `unknown`: The user can conclude that the message was not received by the IC at the attached timestamp. The user can retry the call, or poll the IC using [`read_state`](#http-read-state) requests.

Reviewer 3: > The user can conclude that the message was not received by the IC at the attached timestamp.
This is *wrong*. The message may not be in the certified state tree at that timestamp, but it could be at any later time because the message may indeed be sitting in some artifact pool of some other node and be included in the next block. The message may also have been received earlier and may have been pruned. It's very hard to conclude anything from `unknown` (other than the user may want to poll until expiration to learn whether anything comes up).

> -    `unknown`: The user can conclude that the message was not received by the IC at the attached timestamp. The user can retry the call, or poll the IC using [`read_state`](#http-read-state) requests.

Author: I'll clarify the `unknown` case. But I still believe it is useful always to return the certificate. If the user/agent sees `processing`, or `received` then they have a signal for their message being received by the IC and that they don't need to retry the request.

> -    `unknown`: The user can conclude that the message was not received by the IC at the attached timestamp. The user can retry the call, or poll the IC using [`read_state`](#http-read-state) requests.

Reviewer: > they don't need to retry the request.

the request might still time out even after being "received"

> -    `unknown`: The user can conclude that the message was not received by the IC at the attached timestamp. The user can retry the call, or poll the IC using [`read_state`](#http-read-state) requests.

Reviewer 4: I agree that in the current setup there is not much value in attaching a certificate for processing, received and unknown.

However, I am at odds with our long term plans. It was said that we should clearly not rely on the ingress history to figure out the if an update made it or not.  So in this case I would imagine that long term, if a user wants to check that his ingress message made it, then the user will have to query in certified fashion the canister state. All this is application specific. 

What is not application specific is that most likely the user will need to be able to distinguish between different timestamps. More specifically the user will need to know past what timestamp querying the canister will yield a correct result.

For example if the user queries the canister at time X to verify that an ingress message mutated the state, then X should be definitely bigger than the time at which a certificate was returned when in processing or received 
 (and unknown?!) state.

All this said, if we omit the certificate for those states now and we want to add them later, most likely it won't break much due to the cbor nature of the response.

> -    `unknown`: The user can conclude that the message was not received by the IC at the attached timestamp. The user can retry the call, or poll the IC using [`read_state`](#http-read-state) requests.

Reviewer 5: > What is not application specific is that most likely the user will need to be able to distinguish between different timestamps. More specifically the user will need to know past what timestamp querying the canister will yield a correct result.
Based on the expiry time (which is set by the user), the user can figure out at which point the call must've made it, even without the certificate, though I'm not actually sure where the call will have made it to :) probably the input queue at best, I doubt we guarantee that the message processing actually starts. And I think this would be no better with a certificate. But even if the processing did start at this point, the canister can in principle take an arbitrarily long time to process the message. So the user will in general still need some canister-specific knowledge to figure out at which time the result of a query is definite.

> -    `unknown`: The user can conclude that the message was not received by the IC at the attached timestamp. The user can retry the call, or poll the IC using [`read_state`](#http-read-state) requests.

Author: I changed it to return 202 and an empty body if it's not in a terminal state. This means we will not return a certificate if the call status is not in `replied`, `rejected`, or `done`.

### Request: Call {#http-call}

In order to call a canister, the user makes a POST request to `/api/v3/canister/<effective_canister_id>/call`. The request body consists of an authentication envelope with a `content` map with the following fields:

-   `request_type` (`text`): Always `call`

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call.

-   `arg` (`blob`): Argument to pass to the canister method.

> The HTTP response to this request can have the following responses:

Reviewer: ```suggestion
The HTTP response to this request can have the following forms:
```

> -   200 HTTP status with a non-empty body. This status is returned if the replica received the message.

Reviewer: ```suggestion
-   200 HTTP status with a non-empty body. This status is returned if the replica successfully received the message.
```

>     -   A certificate for the state of the update call is produced, and returned as a CBOR (see [CBOR](#cbor)) map with the fields below. The user should use the certificate to determine the state of the call:

Reviewer: ```suggestion
    -   A certificate for the state of the update call is produced, and returned in a CBOR (see [CBOR](#cbor)) map with the fields specified below. The user should use the certificate to determine the state of the call:
```

        -   `status` (`text`): `"certified_state"`

>         -   `reply` (`blob`):  A certificate (see [Certification](#certification)) with subtrees at `/request_status/<request_id>` and `/time`. See [Request status](#state-tree-request-status) for more details on the request status.

Reviewer: ```suggestion
        -   `reply` (`blob`):  A certificate (see [Certification](#certification)) with subtrees at `/request_status/<request_id>` and `/time`, where `<request_id>` is the [request ID](#request-id) of the update call. See [Request status](#state-tree-request-status) for more details on the request status.
```

>     -   If a non-replicated pre-processing error occurred due to the [canister inspect message](#system-api-inspect-message), then a body with information about the IC specific error encountered is returned. The body is a CBOR map with the following fields:

Reviewer: ```suggestion
    -   If a non-replicated pre-processing error occurred (e.g., due to the [canister inspect message](#system-api-inspect-message)), then a body with information about the IC specific error encountered is returned. The body is a CBOR map with the following fields:
```

>         -   `status` (`text`): `"canister_inspect_message_rejected"`

Reviewer: this is too specific, there can be multiple reasons why a pre-processing error occurs, e.g., the canister is frozen

Author: Changed it to "non_replicated_rejection"

        -   `reject_code` (`nat`): The reject code (see [Reject codes](#reject-codes)).

        -   `reject_message` (`text`): a textual diagnostic message.

        -   `error_code` (`text`): an optional implementation-specific textual error code (see [Error codes](#error-codes)).

> 

Author: I deliberately removed the `202` status response.

I am proposing that we always return a certificate with a `200` HTTP status code if a replica successfully receives a message. It makes the API simpler, and

> 

Author: I deliberately removed the `202` response after a timeout.

I am proposing that we always return a certificate with a 200 HTTP status code in the new API, regardless of certification or timeout. It makes the API much simpler. 

From the user's perspective, the HTTP status code serves no value, as they can't trust the status code alone for what action to perform next. They need to inspect the body and drill down into the certificate from the subnet and c choose the next action based on the state that is certified.

Thus when the agent sees a `200` response with a certificate they can choose the next step of action, (completed, polling /read_state, retrying the HTTP request, etc.) based on the certificate.

> 

Reviewer: > they can't trust the status code alone

sure, but the certificate with "unknown", "received", "processing" doesn't let you conclude anything useful either

> 

Author: `Received` and `processing` lets you conclude that the ingress message made it to the IC.

> 

Reviewer: > Received and processing lets you conclude that the ingress message made it to the IC.

Sure, but here "making it to the IC" doesn't mean "being successfully executed".

> 

Reviewer 4: If the user receives 200 with "non_replicated_rejection" the user can't conclude anything useful as well. 

The thinking here is that the user will get 200 iff it hits the wasm. In other words the IC is a transport for the execution layer in a way.

The replica will maintain the HTTPS connection for the request and will respond once the call status transitions to a terminal state. 

If an implementation specific timeout for the request is reached while the replica waits for the terminal state, then the replica will reply with an empty body and a 202 HTTP status code. In such cases, the user should use [`read_state`](#http-read-state) to determine the status of the call.

### Request: Call {#http-call}

In order to call a canister, the user makes a POST request to `/api/v3/canister/<effective_canister_id>/call`. The request body consists of an authentication envelope with a `content` map with the following fields:

-   `request_type` (`text`): Always `call`

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call.

-   `arg` (`blob`): Argument to pass to the canister method.

The HTTP response to this request can have the following forms:

> -   200 HTTP status with a non-empty body. This status is returned if the canister call completed or was rejected within an implementation-specific timeout.

Reviewer: ```suggestion
-   200 HTTP status with a non-empty body. This status is returned if the canister call completed.
```

Author: The call can be rejected, and we still reply with `200`. Hence the `status` fields below.

Reviewer: > The call can be rejected, and we still reply with 200. Hence the status fields below.

What do you mean?

Author: I mean the ingress filter can fail, and we never submit the ingress message to consensus+p2p, we reply with a `200` status code.

https://github.com/dfinity/interface-spec/pull/265/files/d6edfe5ab5d10c2b08e0d6267fe88c224841232e#diff-b9a833d6114611a080aee4cd57618feb45aa044bcc11105d44b155b5bbde00ebR684-R693

Reviewer: > I mean the ingress filter can fail, and we never submit the ingress message to consensus+p2p, we reply with a 200 status code.

This is covered by "the canister call completed" since the call completed in this case (with an error).

-   `request_type` (`text`): Always `sync_call`

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call.

-   `arg` (`blob`): Argument to pass to the canister method.

The HTTP response to this request can have the following responses:

-   200 HTTP status with a non-empty body. The response contains the canister response, which is either a `reply` or a `reject`.
    
    -   If the update call resulted in a (replicated) reply, the response is a CBOR (see [CBOR](#cbor)) map with the following fields:

        -   `status` (`text`): `"replied"`

        -   `reply` (`blob`):  A certificate (see [Certification](#certification)) with subtrees at `/request_status/<request_id>` and `/time`. See [Request status](#state-tree-request-status) for more details on the request status.

>     -   If the update call resulted in a (non-replicated) reject, the response is a CBOR map with the following fields:

Reviewer: This is pretty subtle, a sync_call can be rejected in the replicated mode with a certificate or in the non-replicated mode without a certificate.

-   `method_name` (`text`): Name of the canister method to call.

-   `arg` (`blob`): Argument to pass to the canister method.

The HTTP response to this request can have the following responses:

-   200 HTTP status with a non-empty body. The response contains the canister response, which is either a `reply` or a `reject`.
    
    -   If the update call resulted in a (replicated) reply, the response is a CBOR (see [CBOR](#cbor)) map with the following fields:

        -   `status` (`text`): `"replied"`

        -   `reply` (`blob`):  A certificate (see [Certification](#certification)) with subtrees at `/request_status/<request_id>` and `/time`. See [Request status](#state-tree-request-status) for more details on the request status.

    -   If the update call resulted in a (non-replicated) reject, the response is a CBOR map with the following fields:

        -   `status` (`text`): `"rejected"`

        -   `reject_code` (`nat`): The reject code (see [Reject codes](#reject-codes)).

        -   `reject_message` (`text`): a textual diagnostic message.

        -   `error_code` (`text`): an optional implementation-specific textual error code (see [Error codes](#error-codes)).

> -   202 HTTP status with an empty body. This status is returned as a fallback mechanism for when a response from the canister was not produced in reasonable time. This implies that the request was accepted by the IC for further processing. Users should use [`read_state`](#http-read-state) to determine the status of the call.

Reviewer 4: What will be a "reasonable time "? Should we just use the `ingress_expiry` value. The annoying part is that when the ingress_expiry is hit then we will be returning 202, which I am not sure how intuitive it is.

Reviewer 6: Just my two cents here: AFAIK `ingress_expiry` is usually set to 5 minutes in the future. Connection timeouts for Web Applications happen usually around ~30 seconds (in my experience). So the reasonable time should be well below the 30 seconds mark, otherwise clients will start seeing HTTP 408 / HTTP 504 and the like (instead of the eventually produced 202 status).

:::note

The functionality exposed via the [The IC management canister](#ic-management-canister) can be used this way.

:::

### Request: Sync Call {#http-sync-call}

In order to make an update call to a canister and potentially get a synchronous response, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/sync_call`. The request body consists of an authentication envelope with a `content` map with the following fields:

-   `request_type` (`text`): Always `sync_call`

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call.

-   `arg` (`blob`): Argument to pass to the canister method.

The HTTP response to this request can have the following responses:

-   200 HTTP status with a non-empty body. The response contains the canister response, which is either a `reply` or a `reject`.
    

>     -   If the update call completes within one execution round, or the call expires, the response is a CBOR (see [CBOR](#cbor)) map with the following fields:

Reviewer: ```suggestion
    -   If the update call completes within an implementation-specific timeout, the response is a CBOR (see [CBOR](#cbor)) map with the following fields:
```

        -   `status` (`text`): `"terminal"`

        -   `reply` (`blob`):  A certificate (see [Certification](#certification)) with subtrees at `/request_status/<request_id>` and `/time`. See [Request status](#state-tree-request-status) for more details on the request status.

    -   If the update call resulted in a (non-replicated) reject, the response is a CBOR map with the following fields:

        -   `status` (`text`): `"non_replicated_rejection"`

        -   `reject_code` (`nat`): The reject code (see [Reject codes](#reject-codes)).

        -   `reject_message` (`text`): a textual diagnostic message.

        -   `error_code` (`text`): an optional implementation-specific textual error code (see [Error codes](#error-codes)).

> -   202 HTTP status with an empty body. This status is returned as a fallback mechanism for when the canister call is running for more than one execution round. This implies that the request was accepted by the IC for further processing. Users should use [`read_state`](#http-read-state) to determine the status of the call.

Reviewer: ```suggestion
-   202 HTTP status with an empty body. This status is returned as a fallback mechanism for when the canister call is running for more than an implementation-specific timeout. This implies that the request was accepted by the IC for further processing. Users should use [`read_state`](#http-read-state) to determine the status of the call.
```

:::

### Request: Sync Call {#http-sync-call}

In order to make a synchronous update call to a canister, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/sync_call`. The request body consists of an authentication envelope with a `content` map with the following fields:

> -   `request_type` (`text`): Always `sync-call`

Reviewer: or sync_call

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call

-   `arg` (`blob`): Argument to pass to the canister method

The HTTP response to this request can have the following responses:

-   202 HTTP status with empty body. Implying the request was accepted by the IC for further processing. Users should use [`read_state`](#http-read-state) to determine the status of the call.

-   200 HTTP status with a non-empty body. The response contains the canister response, which is either a `reply` or a `reject`.
    
    -   If the update call resulted in a reply, the response is a CBOR (see [CBOR](#cbor)) map with the following fields:

        -   `reply` (`blob`):  A certificate (see [Certification](#certification)) with the subtree at `/request_status/<request_id>` and `/time`.

>     -  If the update call resulted in a reject, the response is a CBOR (see [CBOR](#cbor)) map with the following fields:

Reviewer: We should not return two different types of CBOR maps without giving the user a way to tell what to try parsing. Please see my suggestion on the PR: let's introduce an enumeration and wrap the two CBOR maps (for reply and reject) as variants of that enumeration.

Author: I added a `status` field to both CBOR maps with the value set to either `rejected` or `replied`. This is the same way we handle reject and reply for Query calls.  https://internetcomputer.org/docs/current/references/ic-interface-spec#http-query

Is this fine by you?

Reviewer: Yes. That makes sense - Thank you for taking inspiration from query calls!

        -   `reject_code` (`nat`): The reject code (see [Reject codes](#reject-codes)).

        -   `reject_message` (`text`): a textual diagnostic message.

        -   `error_code` (`text`): an optional implementation-specific textual error code (see [Error codes](#error-codes)).

-   4xx HTTP status for client errors (e.g. malformed request). Except for 429 HTTP status, retrying the request will likely have the same outcome.

-   5xx HTTP status when the server has encountered an error or is otherwise incapable of performing the request. The request might succeed if retried at a later time.

This request type can *also* be used to call a query method (but not a composite query method). A user may choose to go this way, instead of via the faster and cheaper [Request: Query call](#http-query) below, if they want to get a *certified* response. Note that the canister state will not be changed by sending a call request type for a query method (except for cycle balance change due to message execution).

> See [Request status](#state-tree-request-status) for details on the request status.

Reviewer: Wouldn't this naturally be a comment next to the `reply` case?

:::

### Request: Sync Call {#http-sync-call}

In order to make a synchronous update call to a canister, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/sync_call`. The request body consists of an authentication envelope with a `content` map with the following fields:

-   `request_type` (`text`): Always `sync_call`

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call

-   `arg` (`blob`): Argument to pass to the canister method

The HTTP response to this request can have the following responses:

> -   202 HTTP status with empty body. Implying the request was accepted by the IC for further processing. Users should use [`read_state`](#http-read-state) to determine the status of the call.

Reviewer 2: Nit: This may make more sense as the second entry, both because 202 comes after 200 and because it's the fallback behavior if a response is not produced in reasonable time.

Also, it may be worth explicitly stating that this is the fallback in case a response was not produced in reasonable time.

This request type can *also* be used to call a query method (but not a composite query method). A user may choose to go this way, instead of via the faster and cheaper [Request: Query call](#http-query) below, if they want to get a *certified* response. Note that the canister state will not be changed by sending a call request type for a query method (except for cycle balance change due to message execution).

### Request: Sync Call {#http-sync-call}

> A synchronous update call, or "call and await", is a type of update [call](#http-call) where the replica will wait with replying to the user until the call has been processed and the result has been added to the replicated state. This means the user will receive a certified result of the call, and thus __do not need to poll__ [`read_state`](#http-read-state) to determine the status of the call.

Reviewer: ```suggestion
A synchronous update call, or "call and await", is a type of update [call](#http-call) where the replica will wait with replying to the user until the call has been processed and the result has been added to the certified state. This means the user will receive a certified result of the call, and thus __does not need to poll__ (using [`read_state`](#http-read-state) requests) to determine the status of the call.
```

In order to make a synchronous update call to a canister, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/sync_call`. The request body consists of an authentication envelope with a `content` map with the following fields:

-   `request_type` (`text`): Always `sync-call`

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call

-   `arg` (`blob`): Argument to pass to the canister method

The HTTP response to this request can have the following responses:

> -   201 HTTP status with a non-empty body. Implying the request was processed.

Reviewer: ```suggestion
-   201 HTTP status with a non-empty body. Implying the request was processed.

```

    -   `reject_code` (`nat`): The reject code (see [Reject codes](#reject-codes)).

    -   `reject_message` (`text`): a textual diagnostic message.

    -   `error_code` (`text`): an optional implementation-specific textual error code (see [Error codes](#error-codes)).

-   4xx HTTP status for client errors (e.g. malformed request). Except for 429 HTTP status, retrying the request will likely have the same outcome.

-   5xx HTTP status when the server has encountered an error or is otherwise incapable of performing the request. The request might succeed if retried at a later time.

This request type can *also* be used to call a query method (but not a composite query method). A user may choose to go this way, instead of via the faster and cheaper [Request: Query call](#http-query) below, if they want to get a *certified* response. Note that the canister state will not be changed by sending a call request type for a query method (except for cycle balance change due to message execution).

:::note

The functionality exposed via the [The IC management canister](#ic-management-canister) can be used this way.

:::

> See [The system state tree](#state-tree) for details on the state tree.

Reviewer: This looks like a copy-paste error. Could you please double check?

Author: I added it since we are returning the state of the ingress call. It should be changed to:

```suggestion
See [Request status](#state-tree-request-status) for details on the request status.
```

:::

### Request: Sync Call {#http-sync-call}
A synchronous update call, or "call and await", is a type of update [call](#http-call) where the replica will wait with replying to the user until the call has been processed and the result has been added to the replicated state. This means the user will receive a certified result of the call, and thus __do not need to poll__ [`read_state`](#http-read-state) to determine the status of the call.
In order to make a synchronous update call to a canister, the user makes a POST request to `/api/v2/canister/<effective_canister_id>/sync_call`. The request body consists of an authentication envelope with a `content` map with the following fields:

-   `request_type` (`text`): Always `sync-call`

-   `sender`, `nonce`, `ingress_expiry`: See [Authentication](#authentication)

-   `canister_id` (`blob`): The principal of the canister to call.

-   `method_name` (`text`): Name of the canister method to call

-   `arg` (`blob`): Argument to pass to the canister method

The HTTP response to this request can have the following responses:

> -   202 HTTP status with a non-empty body. Implying the request was processed.

Reviewer: 202 suggests that the request has been accepted for further processing which is not the case here - the status code 200 should be returned

Author: Good point. I changed it to `201` for the cases where a certification is returned. 

We can keep `202` for cases where the user needs to fall back to polling, for example if we time out the requests because no response has been generated within some time threshold.

Reviewer: Why not 200?

Reviewer: > We can keep 202 for cases where the user needs to fall back to polling, for example if we time out the requests because no response has been generated within some time threshold.

That makes sense!

Author: Thanks for the feedback!

>I do not think HTTP 201 is appropriate here. While the request method is POST, it is not true that the request "has led to the creation of a resource".

Can't we argue that response status + Certification that's returned (which is added to the replicated state) is a "creation of a resource"?

I agree that making `200` generic over `reply` and `reject` would make things more consistent with how canister calls work internally. However, reserving `200` for `reject` makes the API of our sync endpoint more consistent with the async endpoint for the user.

I don't have a strong opinion about either option. We can go with a generic `200` if you think it makes more sense.

CC: @rumenov

    -   `response` (`blob`): A certificate (see [Certification](#certification)).

> The returned certificate includes the subtree at `/request_status/<request_id>` and `/time`.

Reviewer: ```suggestion
    The returned certificate includes the subtree at `/request_status/<request_id>` and `/time`.
```

    -   `reject_code` (`nat`): The reject code (see [Reject codes](#reject-codes)).

    -   `reject_message` (`text`): a textual diagnostic message.

    -   `error_code` (`text`): an optional implementation-specific textual error code (see [Error codes](#error-codes)).

-   4xx HTTP status for client errors (e.g. malformed request). Except for 429 HTTP status, retrying the request will likely have the same outcome.

-   5xx HTTP status when the server has encountered an error or is otherwise incapable of performing the request. The request might succeed if retried at a later time.

This request type can *also* be used to call a query method (but not a composite query method). A user may choose to go this way, instead of via the faster and cheaper [Request: Query call](#http-query) below, if they want to get a *certified* response. Note that the canister state will not be changed by sending a call request type for a query method (except for cycle balance change due to message execution).

:::note

> The functionality exposed via the [The IC management canister](#ic-management-canister) can be used this way.

Reviewer: What's the point of this note?

Author: Removed this.

Reviewer: Shouldn't the note stay under the previous "Call" section?

#### API requests

> We distinguish between the *asynchronous* API requests (type `Request`) passed to `/api/v2/…/call` and `/api/v2/…/sync_call`, which may be present in the IC state, and the *synchronous* API requests passed to `/api/v2/…/read_state` and `/api/v2/…/query`, which are only ephemeral.

Reviewer 2: This is slightly confusing: above (and in the API name) we explicitly call `sync_call` "synchronous". Yet here we categorize it as asynchronous.

Is there any possibility to use a different categorization? If not, maybe the endpoint could be named `call_and_await`.

Reviewer: > Yet here we categorize it as asynchronous.

It is asynchronous in the worst-case.

Reviewer: But I agree that it'd make sense to make this more precise and ideally also formally define the response to sync_call on the happy path: a certificate similar to the case of read state request responses.

Author: [Changed ](https://github.com/dfinity/interface-spec/pull/265/commits/07a46146730e44a1f8ffd5a472c63f3116c81d44)

Reviewer 2: SGTM

I'm just wondering whether "call" or "update call" is the most precise term. "Call" seems too generic, since queries are also calls. And "update call" is too specific, as (I believe that) replicated queries are not technically update calls.

Maybe we should use "replicated call", which should cover all ingress messages. Although I suppose that certified queries could also be claimed to be "replicated calls".

That being said, the above is mostly nitpicking on my part. So feel free to ignore.

Reviewer: > replicated queries are not technically update calls.

I believe they are so changed to "update calls".

Based on this abstract notion of the state, we can describe the behavior of the IC. There are three classes of behaviors:

> -   Asynchronous API requests that are submitted via `/api/v2/…/call` and `/api/v2/…/sync_call`. These transitions describe checks that the request must pass to be considered received.

Reviewer 2: Ditto.
